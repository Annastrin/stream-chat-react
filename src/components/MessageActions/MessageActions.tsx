import { usePopper } from 'react-popper';
import React, {
  Dispatch,
  PropsWithChildren,
  SetStateAction,
  useCallback,
  useMemo,
  useState,
} from 'react';

import { MessageActionsBox } from './MessageActionsBox';

import { ActionsIcon as DefaultActionsIcon } from '../Message/icons';
import { isUserMuted } from '../Message/utils';
import { Modal } from '../Modal/Modal';
import { Portal } from '../Modal/Portal';

import { useChatContext } from '../../context/ChatContext';
import { MessageContextValue, useMessageContext } from '../../context/MessageContext';

import type { DefaultStreamChatGenerics } from '../../types/types';

type MessageContextPropsToPick =
  | 'getMessageActions'
  | 'handleDelete'
  | 'handleFlag'
  | 'handleMute'
  | 'handlePin'
  | 'message';

export type MessageActionsProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = Partial<Pick<MessageContextValue<StreamChatGenerics>, MessageContextPropsToPick>> & {
  /** Wrapping element to which the modal with MessageActionsBox can be appended. By default, the wrapping message list container is used */
  actionsBoxModalContainer?: Element;
  /** Wrapping element id to which the modal with MessageActionsBox can be appended. By default, the wrapping message list container id is used */
  actionsBoxModalContainerId?: string;
  /** Custom component rendered on the message actions button */
  ActionsIcon?: React.FunctionComponent;
  /** Custom class that overrides the default class on the message actions wrapper element */
  customWrapperClass?: string;
  inline?: boolean;
  /** Function determines whether the enclosing message was posted me */
  mine?: () => boolean;
  /** Sets the flag that allows to set visibility class on the message options root element */
  setMessageOptionsVisible?: Dispatch<SetStateAction<boolean>>;
};

export const MessageActions = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(
  props: MessageActionsProps<StreamChatGenerics>,
) => {
  const { mutes } = useChatContext<StreamChatGenerics>('MessageActions');
  const {
    customMessageActions,
    getMessageActions: contextGetMessageActions,
    handleDelete: contextHandleDelete,
    handleFlag: contextHandleFlag,
    handleMute: contextHandleMute,
    handlePin: contextHandlePin,
    isMyMessage,
    message: contextMessage,
    messageListContainer,
    messageListContainerId,
    setEditingState,
  } = useMessageContext<StreamChatGenerics>('MessageActions');

  const {
    actionsBoxModalContainer = messageListContainer,
    actionsBoxModalContainerId = messageListContainerId,
    ActionsIcon = DefaultActionsIcon,
    customWrapperClass = '',
    getMessageActions: propGetMessageActions,
    handleDelete: propHandleDelete,
    handleFlag: propHandleFlag,
    handleMute: propHandleMute,
    handlePin: propHandlePin,
    inline,
    message: propMessage,
    mine,
    setMessageOptionsVisible,
  } = props;

  const belongsToMyMessage = useMemo(() => (mine ? mine() : isMyMessage()), [mine, isMyMessage]);

  const [referenceElement, setReferenceElement] = useState<HTMLButtonElement | null>(null);
  const [popperElement, setPopperElement] = useState<HTMLDivElement | null>(null);
  // todo: make the positioning configurable
  const { attributes, styles } = usePopper(referenceElement, popperElement, {
    placement: belongsToMyMessage ? 'left-start' : 'right-start',
  });

  const getMessageActions = propGetMessageActions || contextGetMessageActions;
  const handleDelete = propHandleDelete || contextHandleDelete;
  const handleFlag = propHandleFlag || contextHandleFlag;
  const handleMute = propHandleMute || contextHandleMute;
  const handlePin = propHandlePin || contextHandlePin;
  const message = propMessage || contextMessage;

  const [actionsBoxOpen, setActionsBoxOpen] = useState(false);

  const isMuted = useCallback(() => isUserMuted(message, mutes), [message, mutes]);

  const messageActions = getMessageActions();

  const closeActionsBox = () => {
    setMessageOptionsVisible?.(false);
    setActionsBoxOpen(false);
  };

  const openActionsBox = (event: React.BaseSyntheticEvent) => {
    event.stopPropagation();
    setMessageOptionsVisible?.(true);
    setActionsBoxOpen(true);
  };

  if (!messageActions.length && !customMessageActions) return null;

  return (
    <>
      {actionsBoxOpen && (
        <Portal container={actionsBoxModalContainer} containerId={actionsBoxModalContainerId}>
          <Modal
            className={{ overlay: 'str-chat__message-actions-box--modal-overlay' }}
            hideCloseButton
            innerContainerProps={{
              style: styles.popper,
              ...attributes.popper,
              onClick: closeActionsBox,
            }}
            onClose={closeActionsBox}
            open={actionsBoxOpen}
            setInnerContainer={setPopperElement}
          >
            <MessageActionsBox
              getMessageActions={getMessageActions}
              handleDelete={handleDelete}
              handleEdit={setEditingState}
              handleFlag={handleFlag}
              handleMute={handleMute}
              handlePin={handlePin}
              isUserMuted={isMuted}
              mine={belongsToMyMessage}
            />
          </Modal>
        </Portal>
      )}
      <MessageActionsWrapper customWrapperClass={customWrapperClass} inline={inline}>
        <button
          aria-expanded={actionsBoxOpen}
          aria-haspopup='true'
          aria-label='Open Message Actions Menu'
          className='str-chat__message-actions-box-button'
          data-testid='message-actions-button'
          onClick={openActionsBox}
          ref={setReferenceElement}
        >
          <ActionsIcon className='str-chat__message-action-icon' />
        </button>
      </MessageActionsWrapper>
    </>
  );
};

export type MessageActionsWrapperProps = {
  customWrapperClass?: string;
  inline?: boolean;
};

const MessageActionsWrapper = (props: PropsWithChildren<MessageActionsWrapperProps>) => {
  const { children, customWrapperClass, inline } = props;

  const defaultWrapperClass = `
  str-chat__message-simple__actions__action
  str-chat__message-simple__actions__action--options
  str-chat__message-actions-container`;

  const wrapperClass = customWrapperClass || defaultWrapperClass;

  const wrapperProps = {
    className: wrapperClass,
    'data-testid': 'message-actions',
  };

  if (inline) return <span {...wrapperProps}>{children}</span>;

  return <div {...wrapperProps}>{children}</div>;
};
