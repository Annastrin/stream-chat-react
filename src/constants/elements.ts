export const DEFAULT_THREAD_MSG_LIST_CONTAINER_ID = 'str-chat__thread-list-container';
export const DEFAULT_MAIN_MSG_LIST_CONTAINER_ID = 'str-chat__main-list-container';
