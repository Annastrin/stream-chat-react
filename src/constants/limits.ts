export const DEFAULT_CHANNEL_PAGE_SIZE = 25;
export const DEFAULT_THREAD_PAGE_SIZE = 50;
