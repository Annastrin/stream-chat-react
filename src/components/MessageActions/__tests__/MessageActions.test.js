import React, { useState } from 'react';
import '@testing-library/jest-dom';
import testRenderer from 'react-test-renderer';
import { act, cleanup, fireEvent, render, screen, waitFor } from '@testing-library/react';

import { MessageActions } from '../MessageActions';

import { DEFAULT_MAIN_MSG_LIST_CONTAINER_ID } from '../../../constants/elements';

import { ChannelActionProvider } from '../../../context/ChannelActionContext';
import { ChannelStateProvider } from '../../../context/ChannelStateContext';
import { ChatProvider } from '../../../context/ChatContext';
import { MessageProvider } from '../../../context/MessageContext';
import { TranslationProvider } from '../../../context/TranslationContext';

import { generateMessage, getTestClient } from '../../../mock-builders';

const wrapperMock = document.createElement('div');
jest.spyOn(wrapperMock, 'addEventListener');

const message = generateMessage();

const defaultProps = {
  getMessageActions: () => ['flag', 'mute'],
  handleDelete: () => {},
  handleFlag: () => {},
  handleMute: () => {},
  handlePin: () => {},
  message,
};

const messageContextValue = {
  getMessageActions: () => ['delete', 'edit', 'flag', 'mute', 'pin', 'react', 'reply'],
  handleDelete: () => {},
  handleFlag: () => {},
  handleMute: () => {},
  handlePin: () => {},
  isMyMessage: () => false,
  message,
  messageListContainerId: DEFAULT_MAIN_MSG_LIST_CONTAINER_ID,
  setEditingState: () => {},
};

const chatClient = getTestClient();

const ContenxtProviders = ({ children }) => (
  <ChatProvider value={{ client: chatClient }}>
    <ChannelStateProvider value={{}}>
      <ChannelActionProvider value={{}}>
        <TranslationProvider value={{ t: (key) => key }}>
          <MessageProvider value={{ ...messageContextValue }}>{children}</MessageProvider>
        </TranslationProvider>
      </ChannelActionProvider>
    </ChannelStateProvider>
  </ChatProvider>
);

function renderMessageActions(customProps, renderer = render) {
  return renderer(
    <ContenxtProviders>
      <div id={DEFAULT_MAIN_MSG_LIST_CONTAINER_ID}>
        <MessageActions {...defaultProps} {...customProps} />
      </div>
    </ContenxtProviders>,
  );
}

const messageActionsButtonTestId = 'message-actions-button';
const messageActionsBoxTestId = 'message-actions-box';
const modalOverlayTestId = 'modal-overlay';

const clickActionsButton = () =>
  act(async () => {
    await fireEvent.click(screen.getByTestId(messageActionsButtonTestId));
  });

describe('<MessageActions /> component', () => {
  afterEach(cleanup);
  beforeEach(jest.clearAllMocks);

  it('should render correctly', () => {
    const tree = renderMessageActions({}, testRenderer.create);
    expect(tree.toJSON()).toMatchInlineSnapshot(`
      <div
        id="str-chat__main-list-container"
      >
        <div
          className="
        str-chat__message-simple__actions__action
        str-chat__message-simple__actions__action--options
        str-chat__message-actions-container"
          data-testid="message-actions"
        >
          <button
            aria-expanded={false}
            aria-haspopup="true"
            aria-label="Open Message Actions Menu"
            className="str-chat__message-actions-box-button"
            data-testid="message-actions-button"
            onClick={[Function]}
          >
            <svg
              className="str-chat__message-action-icon"
              height="4"
              viewBox="0 0 11 4"
              width="11"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.5 3a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm4 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm4 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z"
                fillRule="nonzero"
              />
            </svg>
          </button>
        </div>
      </div>
    `);
  });

  it('should not return anything if message has no actions available', () => {
    const { queryByTestId } = renderMessageActions({
      getMessageActions: () => [],
    });
    // eslint-disable-next-line jest-dom/prefer-in-document
    expect(queryByTestId(messageActionsButtonTestId)).toBeNull();
  });

  it('should open message actions box on click', async () => {
    renderMessageActions();
    await waitFor(() => {
      expect(screen.queryByTestId(messageActionsBoxTestId)).not.toBeInTheDocument();
    });
    await clickActionsButton();
    await waitFor(() => {
      expect(screen.queryByTestId(messageActionsBoxTestId)).toBeInTheDocument();
    });
  });

  it('should close message actions box on icon click if already opened', async () => {
    renderMessageActions();
    expect(screen.queryByTestId(messageActionsBoxTestId)).not.toBeInTheDocument();
    await clickActionsButton();
    await waitFor(() => {
      expect(screen.queryByTestId(messageActionsBoxTestId)).toBeInTheDocument();
    });
    await act(async () => {
      await fireEvent.click(screen.getByTestId(modalOverlayTestId));
    });
    expect(screen.queryByTestId(messageActionsBoxTestId)).not.toBeInTheDocument();
  });

  it('should not close message actions box when user clicks outside the overlay if it is already opened', async () => {
    renderMessageActions();
    await clickActionsButton();
    await waitFor(() => {
      expect(screen.queryByTestId(messageActionsBoxTestId)).toBeInTheDocument();
    });
    await act(async () => {
      await fireEvent.click(document);
    });
    await waitFor(() => {
      expect(screen.queryByTestId(messageActionsBoxTestId)).toBeInTheDocument();
    });
  });

  it('should close message actions box when user presses Escape key', async () => {
    renderMessageActions();
    await clickActionsButton();
    await waitFor(() => {
      expect(screen.queryByTestId(messageActionsBoxTestId)).toBeInTheDocument();
    });
    await act(async () => {
      await fireEvent.keyDown(document, { charCode: 27, code: 'Escape', key: 'Escape' });
    });
    await waitFor(() => {
      expect(screen.queryByTestId(messageActionsBoxTestId)).not.toBeInTheDocument();
    });
  });

  it('should not close actions box open on mouseleave if container ref provided', async () => {
    const customProps = {
      messageWrapperRef: { current: wrapperMock },
    };
    renderMessageActions(customProps);
    await clickActionsButton();
    await waitFor(() => {
      expect(screen.queryByTestId(messageActionsBoxTestId)).toBeInTheDocument();
    });
    await act(async () => {
      await fireEvent.mouseLeave(customProps.messageWrapperRef.current);
    });
    await waitFor(() => {
      expect(screen.queryByTestId(messageActionsBoxTestId)).toBeInTheDocument();
    });
  });

  it('should render with a custom wrapper class when one is set', () => {
    const tree = renderMessageActions(
      {
        customWrapperClass: 'custom-wrapper-class',
      },
      testRenderer.create,
    );
    expect(tree.toJSON()).toMatchInlineSnapshot(`
      <div
        id="str-chat__main-list-container"
      >
        <div
          className="custom-wrapper-class"
          data-testid="message-actions"
        >
          <button
            aria-expanded={false}
            aria-haspopup="true"
            aria-label="Open Message Actions Menu"
            className="str-chat__message-actions-box-button"
            data-testid="message-actions-button"
            onClick={[Function]}
          >
            <svg
              className="str-chat__message-action-icon"
              height="4"
              viewBox="0 0 11 4"
              width="11"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.5 3a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm4 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm4 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z"
                fillRule="nonzero"
              />
            </svg>
          </button>
        </div>
      </div>
    `);
  });

  it('should render with an inline element wrapper when inline set', () => {
    const tree = renderMessageActions(
      {
        inline: true,
      },
      testRenderer.create,
    );
    expect(tree.toJSON()).toMatchInlineSnapshot(`
      <div
        id="str-chat__main-list-container"
      >
        <span
          className="
        str-chat__message-simple__actions__action
        str-chat__message-simple__actions__action--options
        str-chat__message-actions-container"
          data-testid="message-actions"
        >
          <button
            aria-expanded={false}
            aria-haspopup="true"
            aria-label="Open Message Actions Menu"
            className="str-chat__message-actions-box-button"
            data-testid="message-actions-button"
            onClick={[Function]}
          >
            <svg
              className="str-chat__message-action-icon"
              height="4"
              viewBox="0 0 11 4"
              width="11"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.5 3a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm4 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm4 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z"
                fillRule="nonzero"
              />
            </svg>
          </button>
        </span>
      </div>
    `);
  });

  it('modal is attached to custom container when provided container id', async () => {
    const actionsBoxModalContainerId = 'custom-container-id';
    const { asFragment } = render(
      <ContenxtProviders>
        <div id={actionsBoxModalContainerId} />
        <div id={DEFAULT_MAIN_MSG_LIST_CONTAINER_ID}>
          <MessageActions
            {...defaultProps}
            actionsBoxModalContainerId={actionsBoxModalContainerId}
          />
        </div>
      </ContenxtProviders>,
    );
    await clickActionsButton();
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });
  it('modal is attached to custom container when provided container element', async () => {
    const actionsBoxModalContainerId = 'custom-container-id';

    const MessageActionsWrapper = () => {
      const [element, setElement] = useState(null);
      return (
        <>
          <div id={actionsBoxModalContainerId} ref={setElement} />
          <div id={DEFAULT_MAIN_MSG_LIST_CONTAINER_ID}>
            <MessageActions {...defaultProps} actionsBoxModalContainer={element} />
          </div>
        </>
      );
    };

    const { asFragment } = render(
      <ContenxtProviders>
        <MessageActionsWrapper />
      </ContenxtProviders>,
    );
    await clickActionsButton();
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
