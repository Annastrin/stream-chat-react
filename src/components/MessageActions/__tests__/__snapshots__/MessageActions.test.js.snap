// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageActions /> component modal is attached to custom container when provided container element 1`] = `
<DocumentFragment>
  <div
    id="custom-container-id"
  >
    <div
      class="str-chat__modal str-chat__modal--open str-chat__message-actions-box--modal-overlay"
      data-testid="modal-overlay"
    >
      <div
        class="str-chat__modal__inner str-chat-react__modal__inner"
      >
        <div
          data-popper-escaped="true"
          data-popper-placement="right-start"
          data-popper-reference-hidden="true"
          style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px);"
        >
          <div
            class="str-chat__message-actions-box"
            data-testid="message-actions-box"
          >
            <div
              aria-label="Message Options"
              class="str-chat__message-actions-list"
              role="listbox"
            >
              <button
                aria-selected="false"
                class="str-chat__message-actions-list-item str-chat__message-actions-list-item-button"
                role="option"
              >
                Flag
              </button>
              <button
                aria-selected="false"
                class="str-chat__message-actions-list-item str-chat__message-actions-list-item-button"
                role="option"
              >
                Mute
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    id="str-chat__main-list-container"
  >
    <div
      class="
  str-chat__message-simple__actions__action
  str-chat__message-simple__actions__action--options
  str-chat__message-actions-container"
      data-testid="message-actions"
    >
      <button
        aria-expanded="true"
        aria-haspopup="true"
        aria-label="Open Message Actions Menu"
        class="str-chat__message-actions-box-button"
        data-testid="message-actions-button"
      >
        <svg
          class="str-chat__message-action-icon"
          height="4"
          viewBox="0 0 11 4"
          width="11"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.5 3a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm4 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm4 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z"
            fill-rule="nonzero"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<MessageActions /> component modal is attached to custom container when provided container id 1`] = `
<DocumentFragment>
  <div
    id="custom-container-id"
  >
    <div
      class="str-chat__modal str-chat__modal--open str-chat__message-actions-box--modal-overlay"
      data-testid="modal-overlay"
    >
      <div
        class="str-chat__modal__inner str-chat-react__modal__inner"
      >
        <div
          data-popper-escaped="true"
          data-popper-placement="right-start"
          data-popper-reference-hidden="true"
          style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px);"
        >
          <div
            class="str-chat__message-actions-box"
            data-testid="message-actions-box"
          >
            <div
              aria-label="Message Options"
              class="str-chat__message-actions-list"
              role="listbox"
            >
              <button
                aria-selected="false"
                class="str-chat__message-actions-list-item str-chat__message-actions-list-item-button"
                role="option"
              >
                Flag
              </button>
              <button
                aria-selected="false"
                class="str-chat__message-actions-list-item str-chat__message-actions-list-item-button"
                role="option"
              >
                Mute
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    id="str-chat__main-list-container"
  >
    <div
      class="
  str-chat__message-simple__actions__action
  str-chat__message-simple__actions__action--options
  str-chat__message-actions-container"
      data-testid="message-actions"
    >
      <button
        aria-expanded="true"
        aria-haspopup="true"
        aria-label="Open Message Actions Menu"
        class="str-chat__message-actions-box-button"
        data-testid="message-actions-button"
      >
        <svg
          class="str-chat__message-action-icon"
          height="4"
          viewBox="0 0 11 4"
          width="11"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.5 3a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm4 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm4 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z"
            fill-rule="nonzero"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
